File: src/PHP/BehavioralPatterns/Mediator/CommunicationHub/Mediators/GroupMediator.php
<?php

namespace BehavioralPatterns\Mediator\CommunicationHub\Mediators;

class GroupMediator implements IMediator
{
    public array $groups = [];

    public function __construct()
    {
        $this->groups = [];
    }

    public function addGroup($group): void
    {
        $this->groups[] = $group;
    }

    public function notify($sender, $senderArgs): void
    {
        foreach ($this->groups as $group) {
            if ($group->doesParticipantExist($sender)) {
                $participants = $group->getParticipants();

                foreach ($participants as $participant) {
                    if ($participant !== $sender) {
                        $participant->receive($sender, $senderArgs);
                    }
                }
            }
        }
    }
}

File: src/Java/BehavioralPatterns/Mediator/CommunicationHub/src/Mediators/GroupMediator.java
package Mediators;

import Participants.Participant;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class GroupMediator implements IMediator {
    private List<Group> groups = new ArrayList<>();

    public void notify(Participant sender, Object senderArgs) {
        List<Participant> receivers = groups.stream()
                .filter(group -> group.doesParticipantExist(sender))
                .flatMap(group -> group.getParticipants().stream())
                .filter(p -> p != sender)
                .distinct()
                .collect(Collectors.toList());

        receivers.forEach(r -> r.receive(sender, senderArgs));
    }

    public List<Group> getGroups() {
        return groups;
    }

    public void setGroups(List<Group> groups) {
        this.groups = groups;
    }
}
